Fullscreen shell example

4tuw2tuv0 opened the pull request:
I think this example only takes fullscreen so I can't run it inside of an already-running Wayland instance (at least, I don't think) so I can't test it, in full, but I think, even if there's anything which still needs finishing, I think there's much more which is properly finished than not. The places where I'm uncertain, I'll leave comments.

I put comments throughout the example to the line number from the original C file so it's easier for new users to compare what happened in the original with the `guile-wlroots` version.

4tuw2tuv0 committed:
Committing the classes and procedure needed to implement the
fullscreen-shell example (I forget whether these were the only
elements written out in the original C file).

Also, create the needed, respective structs in ~(wlroots types)~ module.

4tuw2tuv0 committed:
Also needed by the particular example file we're trying to implement.

The original C procedure requires the user to create undeclared
variables that the procedure will, then, fill in with values.

Obviously, it's not very idiomatic to Guile/Scheme.

So this procedure only asks for ~output~, from the user, and then
returns both values of ~width~ and ~height~, via ~values~.

4tuw2tuv0 committed:
Another procedure the example needs (~wlr_render_texture_with_matrix~).

Need to pull in the ~(bytestructure-class)~ module for
~bytestructure->pointer~ and ~bytestructure~ to handle the
arrays/vectors.

The original C code uses ~const float matrix[static 9]~ so, like the
compiler, we throw an error if the given list is too long (the error
message is similar to what =gcc= would throw but we use "list", in the
error message, rather than "array").

4tuw2tuv0 committed:
This one isn't used in the example but it's so similar (and related)
to the just-committed procedure that we should probably include it,
for the user's sake.

4tuw2tuv0 committed:
wlroots/types/matrix.scm: create file and add ~wlr-matrix-project-box~ procedure

The original C procedure expects the user to create an undefined array
variable to pass to the procedure for the procedure to fill with a
value.

Eschewing this (Guile/Scheme) unidiomatic behavior,
~wlr-matrix-project-box~ has the user give all the other variables,
instead, and returns a list.

The variable-to-set is defined as ~float mat[static 9]~ (and the
~projection~ array similarly) but the procedure only ever returns 9
values (and uses 9, from ~projection~).

So we hard-code the returned list to a length of 9 (the user's not
allowed to pass in a list, for ~projection~, that's shorter than 9, by
the compiler).

Like with ~wlr_render_texture_with_matrix~, we throw an error if
~projection~ is of a shorter length than 9.

4tuw2tuv0 committed:
wlroots/types/transform.scm: create file and add ~wlr-output-transform-invert~ and ~wlr-output-transform-compose~

I'm pretty certain I can't just return, like this (it'll be a
bytestructure); I need to check with Z572.

4tuw2tuv0 committed:
And make sure we can build these modules.

4tuw2tuv0 committed:
example/fullscreen-shell.scm: add example

4tuw2tuv0 committed:
wlroots/types/drm.scm: create file and add ~wlr-drm-create~

Made these when I was trying to create the =output-layers.c= example
before realizing that example only exists on the =master=
branch (right now) and, therefore, was using procedures and structs
which the version of =wlroots= that this project is using doesn't
have, yet.

No reason, I figure, to let work that's been done go to waste so
throwing this in here, as well.

Z572 commented on wlroots/types/fullscreen-shell-v1.scm:
```diff
@@ -0,0 +1,33 @@
+(define-module (wlroots types fullscreen-shell-v1)
```
please remove v1, different version should in same module，user should known what version need, so just import `(wlroots types fullscreen-shell)` and use expect version procedure

Z572 commented on wlroots/types/fullscreen-shell-v1.scm:
```diff
@@ -0,0 +1,33 @@
+            .display-destroy
+            .global
+
+            wlr-fullscreen-shell-v1-present-surface-event
```
should `<wlr-fullscreen-shell-v1-present-surface-event> `, define-wlr-types-class auto add `<` `>`

Z572 commented on wlroots/types/output.scm:
```diff
@@ -197,6 +197,16 @@
+       (bytestructure->pointer width)
+       (bytestructure->pointer height))
+
+    (values (bytestructure-ref width) (bytestructure-ref height))))
```
i think should use `cons`, just get a width is not usually

Z572 commented on wlroots/render/renderer.scm:
```diff
@@ -36,6 +38,17 @@
+  (ffi:int8 "wlr_render_texture_with_matrix" (list '* '* '* ffi:float))
+  (not (zero? (% (unwrap-wlr-renderer r)
+                 (unwrap-wlr-texture texture)
+                 (if (>= (length matrix) 9)
```
i don't think > 9 is usually,maybe just 9.

Z572 commented on wlroots/render/renderer.scm:
```diff
@@ -36,6 +38,17 @@
+                     (bytestructure->pointer
+                       (bytestructure (bs:vector (length matrix) float)
+                                      (list->vector matrix)))
+                   (error "warning: list argument is too small; matrix requires, at least, 9"))
```
don't use `warning: ` begin, and use `srfi-34` `raise` like
`(raise (condition
                (&message
                 (message  "Not a 9 length vector or list"))))`

Z572 commented on wlroots/render/renderer.scm:
```diff
@@ -36,6 +38,17 @@
+                 (if (>= (length matrix) 9)
+                     (bytestructure->pointer
+                       (bytestructure (bs:vector (length matrix) float)
+                                      (list->vector matrix)))
```
please allow pass vector

Z572 commented on wlroots/render/renderer.scm:
```diff
@@ -38,6 +38,19 @@
+                     (bytestructure->pointer
+                       (bytestructure (bs:vector (length projection) float)
+                                      (list->vector projection)))
+                   (error "warning: list argument is too small; projection requires, at least, 9"))
```
same

Z572 commented on wlroots/render/renderer.scm:
```diff
@@ -38,6 +38,19 @@
+  (ffi:int8 "wlr_render_texture" (list '* '* '* ffi:int ffi:int ffi:float))
+  (not (zero? (% (unwrap-wlr-renderer r)
+                 (unwrap-wlr-texture texture)
+                 (if (>= (length projection) 9)
```
same

Z572 commented on wlroots/types/matrix.scm:
```diff
@@ -0,0 +1,24 @@
+       (unwrap-wlr-box box)
+       (bs:enum->integer %wl-output-transform-enum transform)
+       rotation
+       (if (>= (length projection) max-vector-length)
```
can add a procedure to do convert from 9 length vector or list to bytestructure, and use it.

Z572 commented on wlroots/types/matrix.scm:
```diff
@@ -0,0 +1,24 @@
+       (if (>= (length projection) max-vector-length)
+           (bytestructure->pointer (bytestructure (bs:vector (length projection)
+                                                             float)
+                                                  (list->vector projection)))
```
same

Z572 commented on wlroots/types/transform.scm:
```diff
@@ -0,0 +1,13 @@
+(define-module (wlroots types transform)
```
no need this module, please move procedures to `(wlroots types output)`

Z572 commented on wlroots/types/transform.scm:
```diff
@@ -0,0 +1,13 @@
+  #:use-module (wlroots utils))
+
+(define-wlr-procedure (wlr-output-transform-invert tr)
+  ('* "wlr_output_transform_invert" '(*))
```
use `ffi:int32` like `(ffi:int32 "wlr_output_transform_invert" (list ffi:int32))`

Z572 commented on wlroots/types/transform.scm:
```diff
@@ -0,0 +1,13 @@
+  (% (bs:enum->integer %wl-output-transform-enum tr)))
+
+(define-wlr-procedure (wlr-output-transform-compose tr_a tr_b)
+  ('* "wlr_output_transform_compose" '(* *))
```
same

Z572 commented on wlroots/types/transform.scm:
```diff
@@ -0,0 +1,13 @@
+  ('* "wlr_output_transform_invert" '(*))
+  (% (bs:enum->integer %wl-output-transform-enum tr)))
+
+(define-wlr-procedure (wlr-output-transform-compose tr_a tr_b)
```
tr-a tr-b

Z572 commented on wlroots.scm:
```diff
@@ -19,6 +19,7 @@
              (types damage-ring)
              (types data-control)
              (types data-device)
+             (types fullscreen-shell-v1)
```
move this to add scm file's same commit

Z572 commented on example/fullscreen-shell.scm:
```diff
@@ -0,0 +1,522 @@
+;; functions retrieved from each module as comments above the module
+(use-modules (ice-9          curried-definitions)
+             ;; format
+             (ice-9          format)
```
`((ice-9 format) #:select (format))` is better than comment.

Z572 commented on example/fullscreen-shell.scm:
```diff
@@ -0,0 +1,522 @@
+!#
+;;; An implementation of https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/0.16/examples/fullscreen-shell.c?ref_type=heads
+;; functions retrieved from each module as comments above the module
+(use-modules (ice-9          curried-definitions)
```
just one space

Z572 commented on example/fullscreen-shell.scm:
```diff
@@ -0,0 +1,522 @@
+
+(define* (add-listen* obj symbol proc #:key (destroy-when         obj)
+                                            (remove-when-destroy? #t))
+  (let ([listener (make-wl-listener proc)])
```
please don't use `[]`, `()` is good.

Z572 commented on example/fullscreen-shell.scm:
```diff
@@ -0,0 +1,522 @@
+(define ((output-handle-frame fullscreen-output) listener data)
+  ;;; Line 85
+  ;; struct wlr_renderer *renderer = output->server->renderer;
+  (let-values ([(renderer)     (get-server-renderer (get-output-server fullscreen-output))]
```
use `(srfi srfi-71)`'s `let`

Z572 commented on wlroots/types/drm.scm:
```diff
@@ -0,0 +1,28 @@
+;; A stub implementation of Mesa's wl_drm protocol.
+;;
+;; It only implements the minimum necessary for modern clients to behave
+;; properly. In particular, flink handles are left unimplemented.
```
please remove this comment, guile-wlroots is a ffi library

Z572 commented:
you can try it at tty use tmux and ssh.
run `example/fullscreen-shell.scm` at tmux, and open a ssh to connect you computer, use `tmux a` to attach tmux, and look log or send <kbd>C-c</kbd>

Z572 commented:
i get a black screen after run it, is normal?

4tuw2tuv0 committed:
wlroots/types/fullscreen-shell.scm: move module from ~(wlroots types fullscreen-shell-v1)~

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225751653.

4tuw2tuv0 committed:
wlroots/types/fullscreen-shell.scm: fix <wlr-fullscreen-shell-v1-present-surface-event>

Per
https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225752013;
~define-wlr-types-class~ creates classes and classes are opened and
closed by less- and greater-than signs.

4tuw2tuv0 committed:
wlroots/render/renderer.scm and wlroots/types/matrix.scm: fix ~length~ check of matrices

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225754879.

4tuw2tuv0 committed:
wlroots/render/renderer.scm and wlroots/types/matrix.scm: ~raise~ instead of ~error~

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225756536.

4tuw2tuv0 committed:
wlroots/types/output.scm: merge ~(wlroots types transform)~ module into it

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225758319.

Also, addressed
https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225759769,
in the process of moving the procedures over.

4tuw2tuv0 committed:
wlroots/types/output.scm: change pointers to ffi:int32

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225759532.

4tuw2tuv0 committed:
wlroots/types/drm.scm: remove =wlroots= comments

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225761971.

4tuw2tuv0 committed:
wlroots/types/output.scm: return pair, instead of multiple values

Per https://github.com/Z572/guile-wlroots/pull/4#discussion_r1225753567.

4tuw2tuv0 commented:
Hmm; I'm not sure I'm doing the `tmux` thing properly. I started `tmux` in one of the TTYs and then went back to my main TTY and ran a `tmux a`, like you suggested. It connected but running `example/fullscreen-shell.scm` caused it to switch from TTY 1 to the other TTY and I was stuck as before, with the screen unable to process input so not able to do anything.

As to getting a black screen, I believe so? The only comment they've provided new users is "A minimal fullscreen-shell server. It only supports rendering." so I think it's very similar to the `scene-graph` example, just not using the new Scene Graph API.

Z572 commented:
i rebase and pushed at 490e276aa1943447a069c6c4e042614dfb259903, but not push example/fullscreen.scm, need check why black screen, 
wait https://github.com/Z572/guile-wlroots/tree/fullscreen-shell-example-fixed.

Z572 commented:
now add at f47d4bd6bdac5e60e6a278e996d9c84d72f6f610, because `wlr-renderer-end` is in `(when (get-output-surface fullscreen-output) something)`, when move out, success.
